packages/workflow/src/Extensions/ExpressionExtension.ts

```diff
@@ -88,8 +120,213 @@ export const hasNativeMethod = (method: string): boolean => {
  */
 export const extendTransform = (expression: string): { code: string } | undefined => {
 	try {
-		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
-		const ast = parse(expression);
+		const ast = parse(expression, { parser: { parse: parseWithEsprimaNext } }) as types.ASTNode;
+
+		let currentChain = 1;
+
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+		visit(ast, {
+			// Polyfill optional chaining
+			visitChainExpression(path) {
+				this.traverse(path);
+				const chainNumber = currentChain;
+				currentChain += 1;
+
+				// This is to match our fork of tmpl
+				const globalIdentifier = types.builders.identifier(
+					// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+					// @ts-ignore
+					typeof window !== 'object' ? 'global' : 'window',
+				);
+				const undefinedIdentifier = types.builders.identifier('undefined');
+				const cancelIdentifier = types.builders.identifier(`chainCancelToken${chainNumber}`);
+				const valueIdentifier = types.builders.identifier(`chainValue${chainNumber}`);
+				const cancelMemberExpression = types.builders.memberExpression(
+					globalIdentifier,
+					cancelIdentifier,
+				);
+				const valueMemberExpression = types.builders.memberExpression(
+					globalIdentifier,
+					valueIdentifier,
+				);
+
+				const patchedStack: ExpressionKind[] = [];
+
+				const buildCancelCheckWrapper = (node: ExpressionKind): ExpressionKind => {
+					return types.builders.conditionalExpression(
+						types.builders.binaryExpression(
+							'===',
+							cancelMemberExpression,
+							types.builders.booleanLiteral(true),
+						),
+						undefinedIdentifier,
+						node,
+					);
+				};
```

Comment:
Can we define functions like outside of `visitChainExpression()`? To avoid defining them on every run and to slim down this visitor method.

Reply:
They're defined in there to avoid having to pass in a bunch of arguments or create a bunch of expressions multiple times. This is less of an issue with the current implementation and was more of an issue earlier on. I still prefer this way as I think it makes the rest of the code a lot cleaner.

Speaking of defining them on every run. I think we should add a cache (like tmpl does) for expressions because we're performing these transforms every time an extended expression is run. My main concern is the size of the cache, I'm not sure if it'd be a problem on the backend but it might cause issues over long period of time on the frontend. Although we're likely talking in the 0-10MB range over a long session of editing expressions.